### ISSUE TITLE
[java] Typeresolution not working - missing auxclasspath

### ISSUE THREAD

**Issue by clsaa** [pmd/pmd#2311]
**Affects PMD Version:6.15.0 ** 

**Rule: Define Rule MySelf**

**Description:ASTType#getTypeDefinition() return null.  I want to get all the ASTType in a java file then get the type's full class name. The code can work in JUnitTest, but can not work in my idea plugin.**

![image](https://user-images.githubusercontent.com/32028545/75097483-34821200-55e6-11ea-8d2b-ccff1247d5d3.png)
![image](https://user-images.githubusercontent.com/32028545/75097503-6bf0be80-55e6-11ea-899d-09a57e4d88c9.png)

**Code Sample demonstrating the issue:**

```
@Slf4j
public class EntityOnlyUseInAggregateExceptForRootRule extends AbstractJavaRule {

    @Override
    public Object visit(ASTClassOrInterfaceDeclaration node, Object data) {
        final List<ASTType> types = node.findDescendantsOfType(ASTType.class);
        if (types.isEmpty()) {
            return super.visit(node, data);
        }

        final BuildingBlockStructure buildBlockStructure = BuildBlockUtils.getBuildBlockStructure(data);
        if (buildBlockStructure == null) {
            return super.visit(node, data);
        }
        final List<Entity> entities = buildBlockStructure.getByClass(Entity.class);
        if (entities.isEmpty()) {
            return super.visit(node, data);
        }
        final Set<String> entitySet = entities.stream().map(e -> e.getLocation().getFullClassName()).collect(Collectors.toSet());
        final List<AggregateRoot> aggregateRoots = buildBlockStructure.getByClass(AggregateRoot.class);
        for (AggregateRoot aggregateRoot : aggregateRoots) {
            final Location location = aggregateRoot.getLocation();
            entitySet.remove(location.getFullClassName());
        }
        for (ASTType type : types) {
            if (type != null && entitySet.contains(type.getTypeDefinition().getType().getName())) {
                ViolationUtils.addViolationWithPrecisePosition(this, type, data);
            }
        }
        return super.visit(node, data);
    }
}

/**
 * @author clsaa
 */
public class Main {
    Person person = new Person();

    public static void main(String[] args) {
        Person person2;
    }
}
```

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by clsaa** [comment]
> You're supposed to provide a classloader to your PMDConfiguration ([setClassLoader](https://javadoc.io/doc/net.sourceforge.pmd/pmd-core/latest/net/sourceforge/pmd/PMDConfiguration.html#setClassLoader(java.lang.ClassLoader)) or prependClasspath), so that PMD can find the Class corresponding to Person.

It's not work

My guess is because when the user edits the code in the IDEA editor, the code does not become bytecode and is not loaded by any class loader.
When I run the unit test, I can get TypeDefination because all the classes have become bytecode at this time and can be loaded by the corresponding class loader.
Is there any other solution for this situation?

**Comment by oowekyala** [comment]
Not currently no. With pmd 7 it may become possible to use the ide caches directly and not need class instances. But this feature is still fairly incomplete.

**Comment by adangel** [comment]
@clsaa You'll need to calculate the classpath, that the projects are using, and configuring this when running PMD. The other IDEA Plugin had the same problem (https://github.com/amitdev/PMD-Intellij/issues/68 - fixed with https://github.com/jborgers/PMD-Intellij/commit/fb3ca9dbbbf19e4727e4050fd29c2ef0059291ce).
If the bytecode is not available, because the projects have not been built yet, consider triggering a build - the ought to be some IDEA APIs to do this.
I don't know, how IDEA works - e.g. whether it does incremental compiling automatically (like Eclipse does) when files are changed. Depending on that, you might see outdated bytecode.

**Comment by adangel** [comment]
Refs #2274 and #2283 
@clsaa How can we improve the documentation in this regards?

**Comment by clsaa** [comment]
@adangel  sorry, I have been busy in the past two days and did not reply in time.

> @clsaa You'll need to calculate the classpath, that the projects are using, and configuring this when running PMD. The other IDEA Plugin had the same problem ([amitdev/PMD-Intellij#68](https://github.com/amitdev/PMD-Intellij/issues/68) - fixed with [jborgers/PMD-Intellij@fb3ca9d](https://github.com/jborgers/PMD-Intellij/commit/fb3ca9dbbbf19e4727e4050fd29c2ef0059291ce)).
> I don't know, how IDEA works - e.g. whether it does incremental compiling automatically (like Eclipse does) when files are changed. Depending on that, you might see outdated bytecode.

**Thank you very much, I solved my problem using the above method.**

As a last resort, I  trigger a build. There will be great performance problems for some monolithic applications. In addition, this is not just an IDEA plugin. The static code inspection module, which used PMD will also as a stuck point in DevOps. 

I think incremental compiling automatically through IDEA is not a very good solution. PMD should provide an interface Used to initialize and update the context of a class and cache them.

Using PMD and the code to check should be low-coupling.

**Comment by clsaa** [comment]
> Refs #2274 and #2283
> @clsaa How can we improve the documentation in this regards?

This document is totally enough as a starter document.
From my perspective, I think there should be a more detailed explanation of PMDConfiguration, SourceCodeProcessor, RuleSetFactory, and AbstractJavaRule.
Now I complete my function by looking at the sourcecode of net.sourceforge.pmd.PMD and the above several classes. The document provides a great entry point for reading  sourcecode and using it.
